src/lib.rs

```diff
@@ -294,17 +344,47 @@ impl<T> Sender<T> {
 
 impl<T> Drop for Sender<T> {
     fn drop(&mut self) {
-        // SAFETY: The reference won't be used after the channel is freed in this method
+        // SAFETY: The receiver only ever frees the channel if we are in the MESSAGE or
+        // DISCONNECTED states. If we are in the MESSAGE state, then we called
+        // mem::forget(self), so we should not be in this function call. If we are in the
+        // DISCONNECTED state, then the receiver either received a MESSAGE so this statement is
+        // unreachable, or was dropped and observed that our side was still alive, and thus didn't
+        // free the channel.
         let channel = unsafe { self.channel_ptr.as_ref() };
 
         // Set the channel state to disconnected and read what state the receiver was in
-        match channel.state.swap(DISCONNECTED, SeqCst) {
+        // ORDERING: we don't need release ordering here since there are no modifications we
+        // need to make visible to other thread, and the Err(RECEIVING) branch handles
+        // synchronization independent of this cmpxchg
+        //
+        // EMPTY ^ 001 = DISCONNECTED
+        // RECEIVING ^ 001 = UNPARKING
+        // DISCONNECTED ^ 001 = EMPTY (invalid), but this state is never observed
+        match channel.state.fetch_xor(0b001, Relaxed) {
             // The receiver has not started waiting, nor is it dropped.
             EMPTY => (),
             // The receiver is waiting. Wake it up so it can detect that the channel disconnected.
-            RECEIVING => unsafe { channel.take_waker() }.unpark(),
+            #[allow(unreachable_code)]
+            RECEIVING => {
+                // See comments in Sender::send
+
+                fence(Acquire);
+
+                #[allow(unused_variables)]
+                let waker = unsafe { channel.take_waker() };
+
+                // We still need release ordering here to make sure our read of the waker happens
+                // before this.
+                channel.state.store(DISCONNECTED, Release);
+
+                waker.unpark();
```

Comment:
Can't the unpark be moved up to before the store if the ordering is release? Won't we also need acquire to guarantee we wake the receiver up after we store `DISCONNECTED`? If true, the same issue exists in the `send` method.

Reply:
Now I don't have any docs to point to per se - but what I have always been told (and this is also how it's currently implemented) is that `park`/`unpark` come with implied `Acquire`/`Release` semantics, respectively. I wish this was more clearly documented, but the best justification I have for this assertion is that this pattern is duplicated in the [example](https://doc.rust-lang.org/std/thread/fn.park.html#examples) in `std`.

Really the docs for park/unpark should guarantee stronger memory ordering. I'm considering filing an issue for that.

Reply:
Upon further investigation I come to a different conclusion. Yes, the current implementation of `unpark` has `Release` ordering, making it impossible for the state store to move to after the unpark. However, the documentation for parking states:

> Notice that being unblocked does not imply any synchronization with someone that unparked this thread

The example you point to even rely on the actual fences in the current implementation, and could deadlock if `park`/`unpark` did not actually synchronize. I was going to write an issue about this discrepancy. But then I found the following, someone beat me to it by little under two weeks: https://github.com/rust-lang/rust/pull/99587. So let's follow that PR and see what the libs team say about the issue. Until then we can rely on the current implementation. But if they claim that the docs are correct, we need to insert stronger orderings here in our code to be on the safe side.

Reply:
Yeah the docs as they read currently do not make that promise, but the example given in the docs is allowed to deadlock according to the guarantees provided. So that PR is going to need to end in *some* change to the documentation, and as you said I think it's reasonable to assume that it'll guarantee some sort of release ordering.

Reply:
I realize that we should not be talking about the properties of `std::thread::Thread::unpark()` so much. This method call is on `ReceiverWaker::unpark()`. And that method does **not** have any release ordering guarantees. This unpark will lead to `Waker::wake` if the receiver is receiving anync.

I think we need to add stronger more explicit orderings here no matter how `Thread::unpark` is implemented or what it guarantees.

Reply:
That should be fixable by changing that store to a swap with `AcqRel` ordering and simply discarding the return value.